src/server/pipeline.py

```diff
@@ -209,8 +214,15 @@ async def perform_handshake(
             raise Exception("No longer accepting handshakes, closing.")
 
         if not global_connections.add(connection):
+            # Don't penalize a duplicate connection.
+            mark_attempted = False
             raise ProtocolError(Err.DUPLICATE_CONNECTION, [False])
 
+        if not connection.is_outbound:
```

Comment:
This pipeline is already way too complicated. If possible, can we do connection management limits elsewhere (like, connected to `_state_changed` callbacks)? This `pipeline.py` file desperately needs to be made simpler, not more confusing and fragile.